### PR TITLE
Filelist-log file to yaml and with checksum, to be used with esm_tests for the old filelists.py

### PR DESCRIPTION
**Pull request by mandresm**
**Note**: this PR is to `release` because we want to be able to create this file with the old `filelists.py` and then compare it to our new `filedicts.py` when we are ready to merge our new super feature.

This PR includes the following:
- The filelists log file now lands in the `log` folders, not in `config`
- The filelists log is now a yaml file with the following nesting:
   ```yaml
    <component>:
        <file_tag>:
            source: foo
            intermediate: foo
            target: foo
            checksum: foo
   ```
- checksum is `false` if the file is not found
- deleted `.backup` folder in `src/esm_runscripts`

**TODO**
- [x] Actual tests in levante
- [x] Add the new file log to `esm_tests` last tested info
- [x] Add `compute_and_log_file_checksums` to all `prepcompute` recipes
- [x] Clean up 

Example of the new file:
```yaml
fesom:
  config:
    checksum: 2aa73d9f1a98c7ecb1e53605dc6962f2
    intermediate: /Users/mandresm/Work/file_log/run_20000101-20000131/config/fesom/namelist.config
    source: /Users/mandresm/Codes/esm_tools/namelists//fesom2/2.0/awicm3/v3.1//namelist.config
    target: /Users/mandresm/Work/file_log/run_20000101-20000131/work/namelist.config
    type: config
  fesom.clock:
    checksum: 5291ed1d422d040fcbd37b90a7d52292
    intermediate: /Users/mandresm/Work/file_log/run_20000101-20000131/config/fesom/fesom.clock
    source: /Users/mandresm/Work//file_log/run_20000101-20000131/config/fesom//fesom.clock
    target: /Users/mandresm/Work/file_log/run_20000101-20000131/work/fesom.clock
    type: config
  forcing:
    checksum: f3545eaac138f8efe24de0ed9598a768
    intermediate: /Users/mandresm/Work/file_log/run_20000101-20000131/config/fesom/namelist.forcing
    source: /Users/mandresm/Codes/esm_tools/namelists//fesom2/2.0/awicm3/v3.1//namelist.forcing
    target: /Users/mandresm/Work/file_log/run_20000101-20000131/work/namelist.forcing
    type: config
  ice:
    checksum: 71435aca9a8fd094447d010f28c59686
    intermediate: /Users/mandresm/Work/file_log/run_20000101-20000131/config/fesom/namelist.ice
    source: /Users/mandresm/Codes/esm_tools/namelists//fesom2/2.0/awicm3/v3.1//namelist.ice
    target: /Users/mandresm/Work/file_log/run_20000101-20000131/work/namelist.ice
    type: config
  io: 
    checksum: db4f8f0aba5a5219ac564dae7e46b17a
    intermediate: /Users/mandresm/Work/file_log/run_20000101-20000131/config/fesom/namelist.io
    source: /Users/mandresm/Codes/esm_tools/namelists//fesom2/2.0/awicm3/v3.1//namelist.io
    target: /Users/mandresm/Work/file_log/run_20000101-20000131/work/namelist.io
    type: config
  oce:
    checksum: 190b292463d96ecb998609cc9760ce70
    intermediate: /Users/mandresm/Work/file_log/run_20000101-20000131/config/fesom/namelist.oce
    source: /Users/mandresm/Codes/esm_tools/namelists//fesom2/2.0/awicm3/v3.1//namelist.oce
    target: /Users/mandresm/Work/file_log/run_20000101-20000131/work/namelist.oce
    type: config
oasis3mct:
  areas:
    checksum: false
    intermediate: /Users/mandresm/Work/file_log/run_20000101-20000131/input/oasis3mct/areas.nc
    source: TO-BE-DEFINED-BY-THE-USER/oasis//cy43r3/TCO95-CORE2//areas.nc
    target: /Users/mandresm/Work/file_log/run_20000101-20000131/work/areas.nc
    type: input
```

### CODE REVIEW

File: src/esm_runscripts/filelists.py [hunk 1]
```diff
@@ -774,65 +775,51 @@ def replace_year_placeholder(config):
...
+            model_config = config[model]
+            model_files = {}
+
+            for file in model_config.get(f"{filetype}_sources", []):
+                try:
+                    checksum = hashlib.md5(
+                        open(model_config[f"{filetype}_targets"][file], "rb"
+                    ).read()).hexdigest()
+                except FileNotFoundError as err:
+                    checksum = False
```
Comment (pgierz) [inline src/esm_runscripts/filelists.py]: I would prefer None here, or empty string.

File: src/esm_runscripts/filelists.py [hunk 2]
```diff
@@ -774,65 +775,51 @@ def replace_year_placeholder(config):
...
-            f"{thisrun_config_dir}/"
-            f"{expid}_{it_coupled_model_name}filelist_{datestamp}"
-        )
+        for filetype in filetypes:
+            model_config = config[model]
+            model_files = {}
+
+            for file in model_config.get(f"{filetype}_sources", []):
+                try:
+                    checksum = hashlib.md5(
```
Comment (pgierz) [inline src/esm_runscripts/filelists.py]: One nice check to have in mind: does this give the same answer as shell md5 tool? (It should, or I will be very confused)

**Review by pgierz — APPROVED** [review]
Looks great Miguel!

File: src/esm_runscripts/filelists.py (hunk 2 above)
Comment (mandresm) [reply src/esm_runscripts/filelists.py]: I tried it for a file and it yields the same result (at least on my Mac for that file)

File: src/esm_runscripts/filelists.py (hunk 1 above)
Comment (mandresm) [reply src/esm_runscripts/filelists.py]: Agreed! Solved.

**Review by nwieters — APPROVED** [review]

**Comment by pgierz** [comment]
Already reviewed. @mandresm, do you want me to look at either of your two todos to get this all the way in?

**Comment by mandresm** [comment]
Hi @pgierz, this one is a bit block by its sister PR for the filedicts, since there are name conventions that need to be adopted in both for this new file to work as a comparable file between the old dile sys and the new file dicts. In principle, I'd rather leave this opened until the next sprint.

**Comment by pgierz** [comment]
Pinging this PR to keep it open and avoid the robots killing it, I like this feature, and we should revisit it when we look at `filedicts` again.

**Comment by mandresm** [comment]
I've done quite some modifications, improved and tested everything in Levante. At the moment, the checksum is slowing down the code because is not parallelized. However, this PR is a requirement for checking that all things in the parallelization PRs work, so what I've done is changing the base to a new `feat/file_movements_parallelization` branch, so that this does not go directly to `release` and starts to slow down everyones code. After we introduce the Dask parallelization with another PR to `feat/file_movements_parallelization` we can develop also the parallelization of the checksums. Then we merge everything as a single block from `feat/file_movements_parallelization` to `release` so that the users get the parallelization and the checksums all at once.

**Comment by pgierz** [comment]
We are getting an error in the basic CI install, which I'll look into after reviewing the actual code.

**Review by pgierz — APPROVED** [review]

File: src/esm_runscripts/filelists.py [hunk 3]
```diff
@@ -823,9 +826,94 @@ def log_used_files(config):
...
+    else:
+        logger.error(
+            f"Checksums of files in ``{target}`` directory types are not yet "
+            "supported. Only files in the ``work`` directory currently supported. "
+        )
+        exit(1)
+
+    checksums = {}
+    for f in file_paths:
+        checksums[f] = hashlib.md5(open(f, "rb").read()).hexdigest()
```
Comment (pgierz) [inline src/esm_runscripts/filelists.py:971]: This is what's slowing it down. I'd think that we can parallelise this for loop, but I am not sure how that will work for gathering in the results. (AFAIK you can't parallel-ly write information into a dictionary.)

**Comment by pgierz** [comment]
> We are getting an error in the basic CI install, which I'll look into after reviewing the actual code.

Fixed

File: src/esm_runscripts/filelists.py (hunk 3 above)
Comment (mandresm) [reply src/esm_runscripts/filelists.py:971]: Indeed, that's the loop that needs to be parallelize. With dask futures, it's pretty straight forward: you make a futures list that contains the md5 processes, each of this processes returns the checksum, then you loop over the futures using the `.result` method (that's your gathering) and then you have the list you need to feed into your dictionary. Once we have the infrastructure for initiating the dask cluster you can take care of this parallelization if you want to learn about it :)

**Comment by mandresm** [comment]
> > We are getting an error in the basic CI install, which I'll look into after reviewing the actual code.
> 
> Fixed

thanks!

**Pull request merged**